File: sdk/go/workflow_session_test.go
package cigar

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"reflect"
	"strings"
	"testing"
)

func workflowDigest(character string) string { return "1220" + strings.Repeat(character, 64) }
func workflowRecord(suffix uint8) string {
	return fmt.Sprintf("01890f47-8e7d-7b42-a1d2-3c4d5e6f78%02x", suffix)
}

func initialWorkflowCycle(t *testing.T, session *WorkflowContextSession) {
	t.Helper()
	steps := []error{
		session.RecordPlanCreated(workflowRecord(1), workflowDigest("a"), workflowDigest("1")),
		session.RecordBundleCompiled(workflowDigest("a"), workflowDigest("1")),
		session.RecordMaterialized(workflowDigest("a"), workflowDigest("2"), workflowDigest("3"), 10),
		session.BeginModelInvocation(workflowRecord(2), workflowDigest("4"), workflowDigest("8")),
		session.RecordModelResult(workflowRecord(2), workflowDigest("5")),
	}
	for _, err := range steps {
		if err != nil {
			t.Fatal(err)
		}
	}
}

func advanceWorkflowTarget(t *testing.T, session *WorkflowContextSession) {
	t.Helper()
	steps := []error{
		session.RecordObservation(workflowDigest("6"), 1),
		session.RecordPlanCreated(workflowRecord(3), workflowDigest("b"), workflowDigest("7")),
		session.RecordBundleCompiled(workflowDigest("b"), workflowDigest("7")),
		session.RecordDeltaCompiled(workflowDigest("a"), workflowDigest("b"), workflowDigest("8")),
		session.RecordDeltaApplied(workflowDigest("a"), workflowDigest("b"), workflowDigest("8")),
	}
	for _, err := range steps {
		if err != nil {
			t.Fatal(err)
		}
	}
}

func TestWorkflowSharedContractInventoryIsExact(t *testing.T) {
	contractBytes, err := os.ReadFile("../workflow-context-session.v1.json")
	if err != nil {
		t.Fatal(err)
	}
	var contract struct {
		SchemaVersion           string   `json:"schema_version"`
		MaximumDeltaChainLength uint16   `json:"maximum_delta_chain_length"`
		MaximumReplayCycles     int      `json:"maximum_replay_cycles"`
		Phases                  []string `json:"phases"`
		ErrorCodes              []string `json:"error_codes"`
		ResumeActions           []struct {
			Action      string  `json:"action"`
			OperationID *string `json:"operation_id"`
		} `json:"resume_actions"`
		Events            []string `json:"events"`
		QuarantineReasons []string `json:"quarantine_reasons"`
		RetryFences       struct {
			ProviderInvocation string `json:"provider_invocation"`
			EffectRetry        string `json:"effect_retry"`
		} `json:"retry_fences"`
		ReplayComparisonDimensions []string `json:"replay_comparison_dimensions"`
		ReplayVerification         string   `json:"replay_verification"`
		Telemetry                  struct {
			MaximumAddedSeries int      `json:"maximum_added_series"`
			LabelPolicy        string   `json:"label_policy"`
			Families           []string `json:"families"`
		} `json:"telemetry"`
	}
	if err := json.Unmarshal(contractBytes, &contract); err != nil {
		t.Fatal(err)
	}
	phases := make([]string, 0, len(WorkflowContextPhases()))
	for _, phase := range WorkflowContextPhases() {
		phases = append(phases, string(phase))
	}
	codes := make([]string, 0, len(WorkflowSessionErrorCodes()))
	for _, code := range WorkflowSessionErrorCodes() {
		codes = append(codes, string(code))
	}
	if contract.SchemaVersion != "cigar.sdk-workflow-context-session.v1" ||
		contract.MaximumDeltaChainLength != MaxWorkflowDeltaChainLength ||
		contract.MaximumReplayCycles != MaxWorkflowReplayCycles ||
		!reflect.DeepEqual(contract.Phases, phases) || !reflect.DeepEqual(contract.ErrorCodes, codes) ||
		len(contract.ResumeActions) != len(WorkflowResumeActions()) ||
		!reflect.DeepEqual(contract.Events, WorkflowSessionEventNames()) ||
		!reflect.DeepEqual(contract.QuarantineReasons, []string{"cancelled", "revoked", "invalidated"}) ||
		contract.RetryFences.ProviderInvocation !=
			"durable_invocation_and_idempotency_key_digest_required_before_call" ||
		contract.RetryFences.EffectRetry !=
			"durable_reconciliation_count_must_advance_before_authorized_for_retry" ||
		!reflect.DeepEqual(contract.ReplayComparisonDimensions, []string{
			"bundle_delta_selection", "materialization", "model_result_identity",
			"tool_effect_decisions", "outcome",
		}) || contract.ReplayVerification != "all_exact_identity_dimensions_must_equal" ||
		contract.Telemetry.MaximumAddedSeries != 17 ||
		contract.Telemetry.LabelPolicy != "single_closed_static_dimension_no_identifiers_or_content" ||
		!reflect.DeepEqual(contract.Telemetry.Families, []string{
			"cigar_workflow_context_cycles_total",
			"cigar_workflow_context_selections_total",
			"cigar_workflow_context_delta_blocks_total",
			"cigar_workflow_context_recoveries_total",
			"cigar_workflow_context_replay_dimensions_total",
			"cigar_workflow_context_replay_verifications_total",
		}) {
		t.Fatalf("shared workflow contract differs: %+v", contract)
	}
	for index, action := range WorkflowResumeActions() {
		entry := contract.ResumeActions[index]
		operationID := action.OperationID()
		if entry.Action != string(action) || entry.OperationID == nil && operationID != "" ||
			entry.OperationID != nil && *entry.OperationID != operationID {
			t.Fatalf("resume action %d differs: %+v action=%s operation=%s", index, entry, action, operationID)
		}
	}
}

func TestWorkflowDeltaChainBoundForcesFullBundleCheckpoint(t *testing.T) {
	session := NewWorkflowContextSession()
	initialWorkflowCycle(t, session)
	base := "a"
	for index, target := range []string{"b", "c", "d", "e", "f", "1", "2", "3"} {
		steps := []error{
			session.RecordObservation(workflowDigest("6"), uint64(index+1)),
			session.RecordPlanCreated(workflowRecord(uint8(index+3)), workflowDigest(target), workflowDigest("7")),
			session.RecordBundleCompiled(workflowDigest(target), workflowDigest("7")),
			session.RecordDeltaCompiled(workflowDigest(base), workflowDigest(target), workflowDigest("8")),
			session.RecordDeltaApplied(workflowDigest(base), workflowDigest(target), workflowDigest("8")),
		}
		for _, err := range steps {
			if err != nil {
				t.Fatal(err)
			}
		}
		base = target
		if index+1 < int(MaxWorkflowDeltaChainLength) {
			steps = []error{
				session.CheckpointCycle(),
				session.RecordMaterialized(workflowDigest(base), workflowDigest("2"), workflowDigest("3"), 10),
				session.BeginModelInvocation(
					workflowRecord(uint8(index+20)), workflowDigest("4"), workflowDigest("8"),
				),
				session.RecordModelResult(workflowRecord(uint8(index+20)), workflowDigest("5")),
			}
			for _, err := range steps {
				if err != nil {
					t.Fatal(err)
				}
			}
		}
	}
	if session.DeltaChainLength() != MaxWorkflowDeltaChainLength {
		t.Fatalf("delta chain did not reach bound: %d", session.DeltaChainLength())
	}
	steps := []error{
		session.CheckpointCycle(),
		session.RecordMaterialized(workflowDigest(base), workflowDigest("2"), workflowDigest("3"), 10),
		session.BeginModelInvocation(workflowRecord(40), workflowDigest("4"), workflowDigest("8")),
		session.RecordModelResult(workflowRecord(40), workflowDigest("5")),
		session.RecordObservation(workflowDigest("6"), 9),
		session.RecordPlanCreated(workflowRecord(41), workflowDigest("4"), workflowDigest("7")),
		session.RecordBundleCompiled(workflowDigest("4"), workflowDigest("7")),
	}
	for _, err := range steps {
		if err != nil {
			t.Fatal(err)
		}
	}
	activeBundle, active := session.ActiveBundleID()
	if session.Phase() != WorkflowPhaseBundleReady || !active || activeBundle != workflowDigest("4") ||
		session.DeltaChainLength() != 0 || session.ResumeAction() != WorkflowActionCheckpoint {
		t.Fatalf("full checkpoint not promoted: %s", session)
	}
}

func TestWorkflowNoEffectCycleReachesVerifiedReplay(t *testing.T) {
	session := NewWorkflowContextSession()
	initialWorkflowCycle(t, session)
	advanceWorkflowTarget(t, session)
	bundleID, active := session.ActiveBundleID()
	if !active || bundleID != workflowDigest("b") || session.DeltaChainLength() != 1 ||
		session.ResumeAction() != WorkflowActionCheckpoint {
		t.Fatalf("unexpected ready session: %s", session)
	}
	steps := []error{
		session.CheckpointCycle(),
		session.Finish(),
	}
	for _, err := range steps {
		if err != nil {
			t.Fatal(err)
		}
	}
	baseline, err := session.ReplayIdentity()
	if err != nil {
		t.Fatal(err)
	}
	exact, err := session.CompareReplay(baseline)
	if err != nil || !exact.ExactMatch {
		t.Fatalf("exact replay did not compare equal: comparison=%+v error=%v", exact, err)
	}
	incoherent, err := session.ReplayIdentity()
	if err != nil || incoherent.Cycles[0].SelectedDelta == nil {
		t.Fatalf("replay delta unavailable: %v", err)
	}
	incoherent.Cycles[0].SelectedDelta.BaseBundleID = workflowDigest("c")
	_, err = session.CompareReplay(incoherent)
	var incoherentError *WorkflowSessionError
	if !errors.As(err, &incoherentError) || incoherentError.Code != WorkflowErrorIdentityMismatch {
		t.Fatalf("incoherent delta base was not rejected: %v", err)
	}
	impossibleEffect, err := session.ReplayIdentity()
	if err != nil {
		t.Fatal(err)
	}
	impossibleEffect.Cycles[0].Effect = &WorkflowEffectReplayIdentity{
		EffectID: workflowRecord(8), IntentDigest: workflowDigest("9"), EffectVersion: 3,
		State: WorkflowEffectSucceeded, AttemptCount: 0, ReconciliationCount: 0,
	}
	_, err = session.CompareReplay(impossibleEffect)
	var invalidEffect *WorkflowSessionError
	if !errors.As(err, &invalidEffect) || invalidEffect.Code != WorkflowErrorInvalidEvent {
		t.Fatalf("impossible terminal effect counts were not rejected: %v", err)
	}
	changed, err := session.ReplayIdentity()
	if err != nil {
		t.Fatal(err)
	}
	changed.Cycles[0].OutcomeDigest = workflowDigest("d")
	comparison, err := session.CompareReplay(changed)
	if err != nil || comparison.Outcome != WorkflowReplayDifferent ||
		comparison.BundleDeltaSelection != WorkflowReplayEqual || comparison.ExactMatch {
		t.Fatalf("unexpected replay difference: comparison=%+v error=%v", comparison, err)
	}
	_, err = session.RecordReplayVerified(workflowDigest("c"), workflowRecord(4), changed)
	var mismatch *WorkflowSessionError
	if !errors.As(err, &mismatch) || mismatch.Code != WorkflowErrorIdentityMismatch ||
		session.Phase() != WorkflowPhaseFinished {
		t.Fatalf("mismatched replay was not rejected atomically: phase=%s error=%v", session.Phase(), err)
	}
	if _, err := session.RecordReplayVerified(workflowDigest("c"), workflowRecord(4), baseline); err != nil {
		t.Fatal(err)
	}
	if session.CompletedTurns() != 1 || session.Phase() != WorkflowPhaseReplayVerified ||
		session.ResumeAction() != WorkflowActionComplete {
		t.Fatalf("unexpected replay session: %s", session)
	}
}

func TestWorkflowAmbiguousRetryRequiresRevalidation(t *testing.T) {
	session := NewWorkflowContextSession()
	effectID := workflowRecord(8)
	initialWorkflowCycle(t, session)
	if err := session.RecordEffectPrepared(
		effectID, workflowDigest("9"), 1, WorkflowEffectPrepared, 0, 0,
	); err != nil {
		t.Fatal(err)
	}
	advanceWorkflowTarget(t, session)
	steps := []error{
		session.RecordEffectRevalidated(workflowDigest("b"), true),
		session.RecordEffectAuthorized(effectID, workflowDigest("9"), 2, WorkflowEffectAuthorized, 0, 0),
		session.RecordEffectRevalidated(workflowDigest("b"), true),
		session.RecordEffectDispatched(effectID, workflowDigest("9"), 3, WorkflowEffectUnknown, 1, 0),
	}
	for _, err := range steps {
		if err != nil {
			t.Fatal(err)
		}
	}
	missingReconciliation := session.RecordEffectObserved(
		effectID, workflowDigest("9"), 4, WorkflowEffectAuthorizedForRetry, 1, 0,
	)
	var missingProof *WorkflowSessionError
	if !errors.As(missingReconciliation, &missingProof) || missingProof.Code != WorkflowErrorInvalidEvent {
		t.Fatalf("retry without reconciliation proof was accepted: %v", missingReconciliation)
	}
	if err := session.RecordEffectObserved(
		effectID, workflowDigest("9"), 4, WorkflowEffectAuthorizedForRetry, 1, 1,
	); err != nil {
		t.Fatal(err)
	}
	err := session.RecordEffectDispatched(effectID, workflowDigest("9"), 5, WorkflowEffectSucceeded, 2, 1)
	var transition *WorkflowSessionError
	if !errors.As(err, &transition) || transition.Code != WorkflowErrorInvalidTransition ||
		session.Phase() != WorkflowPhaseEffectAuthorized {
		t.Fatalf("retry bypass was not rejected atomically: err=%v session=%s", err, session)
	}
	steps = []error{
		session.RecordEffectRevalidated(workflowDigest("b"), true),
		session.RecordEffectDispatched(effectID, workflowDigest("9"), 5, WorkflowEffectSucceeded, 2, 1),
		session.CheckpointCycle(),
	}
	for _, err := range steps {
		if err != nil {
			t.Fatal(err)
		}
	}
}

func TestWorkflowCancellationQuarantinesLateProviderResult(t *testing.T) {
	session := NewWorkflowContextSession()
	steps := []error{
		session.RecordPlanCreated(workflowRecord(1), workflowDigest("a"), workflowDigest("1")),
		session.RecordBundleCompiled(workflowDigest("a"), workflowDigest("1")),
		session.RecordMaterialized(workflowDigest("a"), workflowDigest("2"), workflowDigest("3"), 10),
		session.BeginModelInvocation(workflowRecord(2), workflowDigest("4"), workflowDigest("8")),
		session.QuarantineContext(workflowDigest("a"), WorkflowQuarantineCancelled),
	}
	for _, err := range steps {
		if err != nil {
			t.Fatal(err)
		}
	}
	err := session.RecordModelResult(workflowRecord(2), workflowDigest("5"))
	var transition *WorkflowSessionError
	if !errors.As(err, &transition) || transition.Code != WorkflowErrorInvalidTransition ||
		session.Phase() != WorkflowPhaseQuarantined || session.ResumeAction() != WorkflowActionComplete {
		t.Fatalf("late provider result was not quarantined: err=%v session=%s", err, session)
	}
}

func TestWorkflowFailureIsAtomicAndContentFree(t *testing.T) {
	session := NewWorkflowContextSession()
	err := session.RecordBundleCompiled(workflowDigest("a"), workflowDigest("1"))
	var transition *WorkflowSessionError
	if !errors.As(err, &transition) || transition.Code != WorkflowErrorInvalidTransition ||
		session.Phase() != WorkflowPhaseNew || strings.Contains(session.String(), workflowDigest("a")) {
		t.Fatalf("unexpected failure behavior: err=%v session=%s", err, session)
	}
}

File: sdk/typescript/src/tests/workflow-session.test.ts
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { test } from "node:test";

import {
  MAX_WORKFLOW_DELTA_CHAIN_LENGTH,
  MAX_WORKFLOW_REPLAY_CYCLES,
  WORKFLOW_CONTEXT_PHASES,
  WORKFLOW_QUARANTINE_REASONS,
  WORKFLOW_RESUME_ACTIONS,
  WORKFLOW_SESSION_ERROR_CODES,
  WORKFLOW_SESSION_EVENT_NAMES,
  WorkflowContextSession,
  WorkflowSessionError,
  workflowOperationId,
} from "../index.js";

const digests = Object.fromEntries(
  [..."0123456789abcdef"].map((character) => [character, `1220${character.repeat(64)}`]),
) as Readonly<Record<string, string>>;
const digest = (character: string): string => digests[character]!;
const record = (suffix: number): string => `01890f47-8e7d-7b42-a1d2-3c4d5e6f78${suffix.toString(16).padStart(2, "0")}`;

function initialCycle(session: WorkflowContextSession): void {
  session.recordPlanCreated(record(1), digest("a"), digest("1"));
  session.recordBundleCompiled(digest("a"), digest("1"));
  session.recordMaterialized(digest("a"), digest("2"), digest("3"), 10);
  session.beginModelInvocation(record(2), digest("4"), digest("8"));
  session.recordModelResult(record(2), digest("5"));
}

function advanceTarget(session: WorkflowContextSession): void {
  session.recordObservation(digest("6"), 1n);
  session.recordPlanCreated(record(3), digest("b"), digest("7"));
  session.recordBundleCompiled(digest("b"), digest("7"));
  session.recordDeltaCompiled(digest("a"), digest("b"), digest("8"));
  session.recordDeltaApplied(digest("a"), digest("b"), digest("8"));
}

test("shared workflow contract inventory is exact", () => {
  const contract = JSON.parse(
    readFileSync(new URL("../../../workflow-context-session.v1.json", import.meta.url), "utf8"),
  ) as {
    schema_version: string;
    maximum_delta_chain_length: number;
    maximum_replay_cycles: number;
    phases: string[];
    error_codes: string[];
    resume_actions: { action: string; operation_id: string | null }[];
    events: string[];
    quarantine_reasons: string[];
    retry_fences: {
      provider_invocation: string;
      effect_retry: string;
    };
    replay_comparison_dimensions: string[];
    replay_verification: string;
    telemetry: {
      maximum_added_series: number;
      label_policy: string;
      families: string[];
    };
  };
  assert.equal(contract.schema_version, "cigar.sdk-workflow-context-session.v1");
  assert.equal(contract.maximum_delta_chain_length, MAX_WORKFLOW_DELTA_CHAIN_LENGTH);
  assert.equal(contract.maximum_replay_cycles, MAX_WORKFLOW_REPLAY_CYCLES);
  assert.deepEqual(contract.phases, WORKFLOW_CONTEXT_PHASES);
  assert.deepEqual(contract.error_codes, WORKFLOW_SESSION_ERROR_CODES);
  assert.deepEqual(contract.resume_actions, WORKFLOW_RESUME_ACTIONS.map((action) => ({
    action,
    operation_id: workflowOperationId(action),
  })));
  assert.deepEqual(contract.events, WORKFLOW_SESSION_EVENT_NAMES);
  assert.deepEqual(contract.quarantine_reasons, WORKFLOW_QUARANTINE_REASONS);
  assert.deepEqual(contract.retry_fences, {
    provider_invocation: "durable_invocation_and_idempotency_key_digest_required_before_call",
    effect_retry: "durable_reconciliation_count_must_advance_before_authorized_for_retry",
  });
  assert.deepEqual(contract.replay_comparison_dimensions, [
    "bundle_delta_selection",
    "materialization",
    "model_result_identity",
    "tool_effect_decisions",
    "outcome",
  ]);
  assert.equal(contract.replay_verification, "all_exact_identity_dimensions_must_equal");
  assert.deepEqual(contract.telemetry, {
    maximum_added_series: 17,
    label_policy: "single_closed_static_dimension_no_identifiers_or_content",
    families: [
      "cigar_workflow_context_cycles_total",
      "cigar_workflow_context_selections_total",
      "cigar_workflow_context_delta_blocks_total",
      "cigar_workflow_context_recoveries_total",
      "cigar_workflow_context_replay_dimensions_total",
      "cigar_workflow_context_replay_verifications_total",
    ],
  });
});

test("no-effect cycle reaches verified replay", () => {
  const session = new WorkflowContextSession();
  initialCycle(session);
  advanceTarget(session);
  assert.equal(session.activeBundleId, digest("b"));
  assert.equal(session.deltaChainLength, 1);
  assert.equal(session.resumeAction, "checkpoint");
  session.checkpointCycle();
  session.finish();
  const baseline = session.replayIdentity();
  const exact = session.compareReplay(baseline);
  assert.equal(exact.exactMatch, true);
  const incoherent = {
    cycles: baseline.cycles.map((cycle, index) => index === 0
      ? { ...cycle, selectedDelta: { ...cycle.selectedDelta!, baseBundleId: digest("c") } }
      : cycle),
  };
  assert.throws(
    () => session.compareReplay(incoherent),
    (error: unknown) => error instanceof WorkflowSessionError && error.code === "identity_mismatch",
  );
  const impossibleEffect = {
    cycles: baseline.cycles.map((cycle, index) => index === 0 ? {
      ...cycle,
      effect: {
        effectId: record(8), intentDigest: digest("9"), effectVersion: 3n,
        state: "succeeded" as const, attemptCount: 0, reconciliationCount: 0,
      },
    } : cycle),
  };
  assert.throws(
    () => session.compareReplay(impossibleEffect),
    (error: unknown) => error instanceof WorkflowSessionError && error.code === "invalid_event",
  );
  const changed = {
    cycles: baseline.cycles.map((cycle, index) =>
      index === 0 ? { ...cycle, outcomeDigest: digest("d") } : cycle),
  };
  const comparison = session.compareReplay(changed);
  assert.equal(comparison.outcome, "different");
  assert.equal(comparison.bundleDeltaSelection, "equal");
  assert.throws(
    () => session.recordReplayVerified(digest("c"), record(4), changed),
    (error: unknown) => error instanceof WorkflowSessionError && error.code === "identity_mismatch",
  );
  assert.equal(session.phase, "finished");
  session.recordReplayVerified(digest("c"), record(4), baseline);
  assert.equal(session.completedTurns, 1);
  assert.equal(session.phase, "replay_verified");
  assert.equal(session.resumeAction, "complete");
});

test("delta chain bound forces a full-bundle checkpoint", () => {
  const session = new WorkflowContextSession();
  initialCycle(session);
  let base = "a";
  [..."bcdef123"].forEach((target, index) => {
    session.recordObservation(digest("6"), BigInt(index + 1));
    session.recordPlanCreated(record(index + 3), digest(target), digest("7"));
    session.recordBundleCompiled(digest(target), digest("7"));
    session.recordDeltaCompiled(digest(base), digest(target), digest("8"));
    session.recordDeltaApplied(digest(base), digest(target), digest("8"));
    base = target;
    if (index + 1 < MAX_WORKFLOW_DELTA_CHAIN_LENGTH) {
      session.checkpointCycle();
      session.recordMaterialized(digest(base), digest("2"), digest("3"), 10);
      session.beginModelInvocation(record(index + 20), digest("4"), digest("8"));
      session.recordModelResult(record(index + 20), digest("5"));
    }
  });
  assert.equal(session.deltaChainLength, MAX_WORKFLOW_DELTA_CHAIN_LENGTH);

  session.checkpointCycle();
  session.recordMaterialized(digest(base), digest("2"), digest("3"), 10);
  session.beginModelInvocation(record(40), digest("4"), digest("8"));
  session.recordModelResult(record(40), digest("5"));
  session.recordObservation(digest("6"), 9n);
  session.recordPlanCreated(record(41), digest("4"), digest("7"));
  session.recordBundleCompiled(digest("4"), digest("7"));
  assert.equal(session.phase, "bundle_ready");
  assert.equal(session.activeBundleId, digest("4"));
  assert.equal(session.deltaChainLength, 0);
  assert.equal(session.resumeAction, "checkpoint");
});

test("ambiguous effect retry requires another revalidation", () => {
  const session = new WorkflowContextSession();
  const effectId = record(8);
  initialCycle(session);
  session.recordEffectPrepared(effectId, digest("9"), 1n);
  advanceTarget(session);
  assert.equal(session.resumeAction, "revalidate_context_bundle");
  session.recordEffectRevalidated(digest("b"), true);
  assert.equal(session.phase, "effect_authorization_revalidated");
  session.recordEffectAuthorized(effectId, digest("9"), 2n);
  session.recordEffectRevalidated(digest("b"), true);
  session.recordEffectDispatched(effectId, digest("9"), 3n, "unknown", 1, 0);
  assert.throws(
    () => session.recordEffectObserved(effectId, digest("9"), 4n, "authorized_for_retry", 1, 0),
    (error: unknown) => error instanceof WorkflowSessionError && error.code === "invalid_event",
  );
  session.recordEffectObserved(effectId, digest("9"), 4n, "authorized_for_retry", 1, 1);
  assert.throws(
    () => session.recordEffectDispatched(effectId, digest("9"), 5n, "succeeded", 2, 1),
    (error: unknown) => error instanceof WorkflowSessionError && error.code === "invalid_transition",
  );
  assert.equal(session.phase, "effect_authorized");
  session.recordEffectRevalidated(digest("b"), true);
  session.recordEffectDispatched(effectId, digest("9"), 5n, "succeeded", 2, 1);
  session.checkpointCycle();
});

test("cancellation quarantines a late provider result", () => {
  const session = new WorkflowContextSession();
  session.recordPlanCreated(record(1), digest("a"), digest("1"));
  session.recordBundleCompiled(digest("a"), digest("1"));
  session.recordMaterialized(digest("a"), digest("2"), digest("3"), 10);
  session.beginModelInvocation(record(2), digest("4"), digest("8"));
  session.quarantineContext(digest("a"), "cancelled");
  assert.throws(
    () => session.recordModelResult(record(2), digest("5")),
    (error: unknown) => error instanceof WorkflowSessionError && error.code === "invalid_transition",
  );
  assert.equal(session.phase, "quarantined");
  assert.equal(session.resumeAction, "complete");
});

test("failed transition is atomic and content-free", () => {
  const session = new WorkflowContextSession();
  assert.throws(
    () => session.recordBundleCompiled(digest("a"), digest("1")),
    (error: unknown) => error instanceof WorkflowSessionError && error.code === "invalid_transition",
  );
  assert.equal(session.phase, "new");
  assert.equal(session.toString().includes(digest("a")), false);
});
